GH-22 Create RPM for srcmap-reverse

johnsonw opened the pull request:
Fixes #22.

- Create RPM for srcmap-reverse.

Signed-off-by: Will Johnson <william.c.johnson@intel.com>

jgrund commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,59 @@
+
+Name:       srcmap-reverse
+Version:    3.0.2
+Release:    2%{?dist}
```
Why release 2 instead of 1?

jgrund commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,59 @@
+Name:       srcmap-reverse
+Version:    3.0.2
+Release:    2%{?dist}
+Summary:    Service that reports current supervisor status as JSON.
```
Summary need updating.

jgrund commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,59 @@
+BuildRequires:  nodejs-packaging
+
+%description
+This module will run an http server listening on /var/run/srcmap-reverse.sock. When the server receives
```
an http / a http

brianjmurrell replied:
"a http" rather than "an http"?

jgrund commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,59 @@
+cp systemd-units/srcmap-reverse.service $RPM_BUILD_ROOT/usr/lib/systemd/system/srcmap-reverse.service
+mkdir -p $RPM_BUILD_ROOT/usr/sbin/
+cp dist/srcmap-reverse.js $RPM_BUILD_ROOT/usr/sbin/srcmap-reverse
+cp dist/main.js.map $RPM_BUILD_ROOT/usr/sbin/main.js.map
```
Hrrm, I'm not sure /usr/sbin is the best spot for this. @brianjmurrell What's the right spot for something that is effectively a metadata file?

brianjmurrell replied:
If you cannot execute it, it doesn't belong in a _bin_ directory.  _/usr/lib_ is probably a better candidate if it's a "static" (i.e. users are not expected to modify it).

johnsonw replied:
Yeah I wasn't sure if this was the best spot either. I'll move to /usr/lib. I will need to update srcmap-reverse so that it looks in this spot though.

jgrund commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,59 @@
+
+%files
+%attr(0744,root,root)/usr/sbin/srcmap-reverse
+%attr(0744,root,root)/usr/sbin/main.js.map
```
This file doesn't need to be executable.

johnsonw replied:
Does this look more correct?
```
+%attr(0444,root,root)/usr/sbin/main.js.map
```

brianjmurrell replied:
Yup, that looks fine.

brianjmurrell commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,59 @@
+systemctl disable srcmap-reverse.socket
+
+%changelog
+* Fri Jul 21 2017 William Johnson <william.c.johnson@intel.com> - 3.0.2
```
You should add the release to end of the version here so that it says **3.0.2-1** (when you fix up the _Release: 2_ above to be _Release: 1_).

johnsonw replied:
Sounds good. I will need to make an update for srcmap-reverse to look for the map file in /usr/lib so I will make it 3.0.3-1.

brianjmurrell replied:
+1

brianjmurrell commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,59 @@
+
+%post
+systemctl enable srcmap-reverse.socket
+systemctl start srcmap-reverse.socket
```
You should not alter the state of either of these on upgrades.  If the admin has stopped and/or disabled the service (for whatever reason they might have decided to do so) we should not "undo" that decision of theirs.

brianjmurrell commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,59 @@
+systemctl stop srcmap-reverse.service
+systemctl disable srcmap-reverse.service
+systemctl stop srcmap-reverse.socket
+systemctl disable srcmap-reverse.socket
```
Be careful of the upgrade case here.  `%preun` will be run during a package upgrade.  Here's a good [explanation of how scriptlets operate in the various install/upgrade/removal cases](https://fedoraproject.org/wiki/Packaging:Scriptlets).

Probably also unnecessary to _disable_ the service here in the package removal case since removal will remove the unit file altogether.

johnsonw replied:
Thanks for this reference Brian.

johnsonw commented:
Adding new tar file separately.

brianjmurrell commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,63 @@
+%attr(0744,root,root)/usr/sbin/srcmap-reverse
+%attr(0444,root,root)/usr/lib/main.js.map
+%attr(0744,root,root)/usr/lib/systemd/system/srcmap-reverse.service
+%attr(0744,root,root)/usr/lib/systemd/system/srcmap-reverse.socket
```
Systemd unit files should not be executable.  `%attr(0644,root,root)` would be more appropriate.

brianjmurrell commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,63 @@
+fi
+
+%changelog
+* Fri Jul 21 2017 William Johnson <william.c.johnson@intel.com> - 3.0.3
```
Don't forget your -1 on the end of the version

johnsonw replied:
I think Joe said tito takes care of this. Joe, is that accurate or do I need to add the -1 on the end?

jgrund replied:
I was thinking of adding a new tag, not modifying an existing one. I think tito adds new tags with releases, but now I'm not sure.

johnsonw replied:
Ok, for now I will add it to be on the safe side.

brianjmurrell replied:
`tito` will add a new `%changelog` entry if you don't ask it not to.

brianjmurrell commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -39,10 +39,10 @@ cp dist/main.js.map $RPM_BUILD_ROOT/usr/lib/main.js.map
 
 %files
-%attr(0744,root,root)/usr/sbin/srcmap-reverse
+%attr(0644,root,root)/usr/sbin/srcmap-reverse
```
This one should be 755 since it's an executable.

brianjmurrell commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,63 @@
+cp systemd-units/srcmap-reverse.socket $RPM_BUILD_ROOT/usr/lib/systemd/system/srcmap-reverse.socket
+cp systemd-units/srcmap-reverse.service $RPM_BUILD_ROOT/usr/lib/systemd/system/srcmap-reverse.service
+mkdir -p $RPM_BUILD_ROOT/usr/sbin/
+cp dist/srcmap-reverse.js $RPM_BUILD_ROOT/usr/sbin/srcmap-reverse
```
Sorry.  I didn't realise that this `srcmap-reverse[.js]` is not directly executable but rather is a node JS program.

Given that it's not directly executable it really shouldn't be in a _bindir_.

Perhaps we should put this in `/usr/share/chroma-manager/`.

brianjmurrell commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,63 @@
+cp systemd-units/srcmap-reverse.service $RPM_BUILD_ROOT/usr/lib/systemd/system/srcmap-reverse.service
+mkdir -p $RPM_BUILD_ROOT/usr/sbin/
+cp dist/srcmap-reverse.js $RPM_BUILD_ROOT/usr/sbin/srcmap-reverse
+cp dist/main.js.map $RPM_BUILD_ROOT/usr/lib/main.js.map
```
This should also probably be in `/usr/share/chroma-manager/` since it's quite specific to IML.

jgrund commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -0,0 +1,64 @@
+  systemctl disable iml-srcmap-reverse.service
+  systemctl stop iml-srcmap-reverse.socket
+  systemctl disable iml-srcmap-reverse.socket
+  rm /var/run/iml-srcmap-reverse.sock
```
Does this not happen with a stop and disable?

johnsonw replied:
I tested on my vm and it did not remove the sock file.

jgrund commented on srcmap-reverse/iml-srcmap-reverse.service:
```diff
@@ -0,0 +1,11 @@
+[Service]
+Restart=always
+Environment=NODE_ENV=production
+Environment=SRCMAP_REVERSE_FD=3
```
Assuming this is used in the module?

johnsonw replied:
Yes, I set it as an environment variable for when the bundle is loaded. It then sets the server to listen with:
```
server.listen({ fd: parseInt(process.env.SRCMAP_REVERSE_FD, 10) });
```

brianjmurrell commented on srcmap-reverse/srcmap-reverse.spec:
```diff
@@ -28,36 +30,35 @@ by line.
-cp dist/main.js.map $RPM_BUILD_ROOT/usr/lib/main.js.map
+mkdir -p %{buildroot}/usr/lib/systemd/system/
+cp %{_sourcedir}/iml-srcmap-reverse.socket %{buildroot}/usr/lib/systemd/system/iml-srcmap-reverse.socket
+cp %{_sourcedir}/iml-srcmap-reverse.service %{buildroot}/usr/lib/systemd/system/iml-srcmap-reverse.service
```
It's better to refer to these as `%{SOURCE[0-n]}` since you then don't have to change a file name in multiple places.

johnsonw replied:
Nice, i'll update this and verify it on my vm.

johnsonw commented:
I rebased on master and updated a few hard coded areas with macros. I tested the updates on my vm and both the installation and removal of the rpm are successful. Integration tests pass after installing the rpm.

brianjmurrell commented on srcmap-reverse/iml-srcmap-reverse.service:
```diff
@@ -0,0 +1,11 @@
+RefuseManualStart=true
+
+[Service]
+Restart=always
```
Is this correct for a service launched from a `.socket` unit?  This seems counter-intuitive for a service that starts on receipt of a network request for it.

jgrund replied:
Yeah, there is really no reason to have this here. This is in the other services as well, so I'll need to update them.

jgrund replied:
The default is `no` so we can just omit this line.

johnsonw replied:
My understanding is that the service should always restart but only become active when the socket is called. This file was based off of supervisor-status: https://github.com/intel-hpdd/supervisor-status/blob/master/systemd-units/supervisor-status.service.

One example I can think of (off the top of my head) is if someone manually tried to send data to the service in a non-json format. This could potentially cause the program to crash; But we would want the service to kick right back up and be ready for a new request that is in the correct format.

jgrund replied:
Right, but the service *should* start again, but only when someone writes to the socket.

jgrund replied:
Example: 

![screen shot 2017-07-26 at 4 43 35 pm](https://user-images.githubusercontent.com/458717/28642852-bf34adc8-7221-11e7-8b9b-f63204931a42.png)
